### ISSUE TITLE
Anon lifetime in `impl Trait` no longer suggests adding a lifetime parameter

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#100615]
Given `fn foo(_: impl Iterator<Item = &u32>) {}`.

In 1.63:

```
error[E0106]: missing lifetime specifier
 --> src/lib.rs:1:32
  |
1 | fn foo(_: impl Iterator<Item = &u32>) {}
  |                                ^ expected named lifetime parameter
  |
help: consider introducing a named lifetime parameter
  |
1 | fn foo<'a>(_: impl Iterator<Item = &'a u32>) {}
  |       ++++                          ++
```

In 1.64:

```
error[E0658]: anonymous lifetimes in `impl Trait` are unstable
 --> src/lib.rs:1:32
  |
1 | fn foo(_: impl Iterator<Item = &u32>) {}
  |                                ^
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23100615.20Anon.20lifetime.20in.20.60impl.20Trait.60.20no.20longer.20suggests.20ad.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by pnkfelix** [comment]
self-assigning to ensure we record a cargo-bisect to the injecting PR

**Comment by apiraino** [comment]
I went ahead and attempted a bisection. The breadcrumbs seem to lead to rollup of #99231 with possibly #97720 being the patchset where this issue originates cc @cjgillot ?

searched nightlies: from nightly-2022-07-13 to nightly-2022-07-15
regressed nightly: nightly-2022-07-15
searched commit range: https://github.com/rust-lang/rust/compare/87588a2afd9ca903366f0deaf84d805f34469384...c2f428d2f3340a0e7d995f4726223db91b93704c
regressed commit: https://github.com/rust-lang/rust/commit/f1a8854f9be2e5cad764d630a53d26c7b72f8162

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.3</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start 2022-07-13 --end 2022-07-15 --preserve 
```
</details>

**Comment by spastorino** [comment]
I'm not sure what would be wrong after @cjgillot change. I think this is exactly the support Camille has added. If you use `#![feature(anonymous_lifetime_in_impl_trait)]` the example compiles and if you don't the error points to the anonymous lifetime and tells you that you would need the feature flag to support such case.

**Comment by cjgillot** [comment]
Added an explanation in https://github.com/rust-lang/rust/issues/85417#issuecomment-1234812068.
TLDR: I proposed to allow this syntax, we made a feature gate instead, and I forgot to put back the suggestion.

**Comment by apiraino** [comment]
Visited during last weeks' T-compiler [meeting on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-09-08/near/297800045), relaying Felix's comment here for visibility:

@cjgillot are you planning to put back the suggestion? Alternatively, do you think that might be a nice small task for a new contributor to pick up?

**Comment by cjgillot** [comment]
This should be an easy enough task for a new contributor.  Summary instructions:
- the feature-gate error is emitted from `rustc_resolve::late::lifetimes::resolve_lifetime_ref`;
- the generics of the enclosing item can be accessed by `tcx.hir().get_generics(lifetime_ref.hir_id.owner)`;
- if that generics struct is `Some`, this means the item can have generics, not that it actually has;
- the contained value is a `hir::Generics`, which should be analysed to know how to suggest: `span` contains the full `<...>` span, and `params` shows how many of them each with their own span;
- I recommend suggesting the additional lifetime parameter as the first one, to avoid questions about the order of lifetime/type/const parameters.

**Comment by Stoozy** [comment]
Hello, I would like to work on this. So far I've been able to get this kind of output
```
error[E0106]: missing lifetime specifier
 --> test.rs:1:32
  |
1 | fn foo(_: impl Iterator<Item = & u32>) {}
  |                                ^ expected named lifetime parameter
  |
help: consider introducing a named lifetime parameter
 --> test.rs:1:7
  |
1 | fn foo(_: impl Iterator<Item = & u32>) {}
  |       ^
```
with the input found in the original message. 

 I am looping over all generic params and checking if the full `<...>` span doesn't contain the `span` associated with the parameter, at which point I try to suggest the introduction of the named lifetime parameter. 
 
 I'm looking for a few pointers on:
 - If my logic actually makes sense
 - what other suggestions `rustc_resolve::late::lifetimes::resolve_lifetime_ref` should handle.
 - how to properly emit the errors/suggestions (I see that late/diagnostics.rs also does similar suggestions but I'm unsure if/how I should use it)

**Comment by Rageking8** [comment]
@Stoozy Another friendly tip: when you want to undertake an issue, you should claim the issue first to prevent duplicated effort (Write the command shown below in the target issue as a comment to assign urself). Thanks.

![Screenshot_2022-09-26-17-40-49-17_320a9a695de7cdce83ed5281148d6f19.jpg](https://user-images.githubusercontent.com/106309953/192245607-ba41d65f-7671-45b0-b602-e28863ef9d0e.jpg)

**Comment by estebank** [comment]
@Stoozy can you call `LateResolutionVisitor::suggest_introducing_lifetime` (or split part of its body to make it appropriate for this case too) and reuse that logic?

**Comment by BGR360** [comment]
Closed by #102323? Or is there more to do?

**Comment by cjgillot** [comment]
Yes, closed by https://github.com/rust-lang/rust/pull/102323, and https://github.com/rust-lang/rust/pull/104048 will handle some suggestion corner cases.
